tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -5,149 +5,175 @@
 //      modified on Mon Jun 19 14:28:04 PDT 2000 by yuanyu
 package tlc2.util;
 
+import java.io.File;
+import java.io.IOException;
+
 /**
  * A <code>BufferedRandomAccessFile</code> is like a
  * <code>RandomAccessFile</code>, but it uses a private buffer
- * so that most operations do not require a disk access.<P>
- * 
- * Note: The operations on this class are unmonitored.
- * The code was written by Allan Heydon
+ * so that most operations do not require a disk access.
+ *
+ * <p>This class is not thread safe.
```

Comment:
Since we're specifying and re-examining the properties of this class, it's maybe interesting that it also has slightly weaker guarantees across executions than RAF.  With both RAF and BRAF, should the program crash in the middle of a write to disk, then the value on disk may be corrupted.

However, unique to BRAF, we also lose causality between the state of the file and future side effects.  With RAF, if you do `write()` then `f()`, and you can confirm that `f()` took place, then you also know that the `write()` took place.  With BRAF, it's possible that the `write()` call did not flush before a crash that occurred after `f()`.  So confirming that `f()` took place does not imply that the `write()` call did.

I wouldn't expect this has implications for tlc, but it is interesting to call out as another place where the semantics are weaker.

Reply:
Good point!